File: KF.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Kalman Filter implementation
%
% Ian Loefgren
% Last modified: 11.11.2018
%
% Implements linear kalman filter
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

classdef KF < handle
    properties
        F
        G
        H
        M
        Q
        R
        x
        P
    end
    methods
        function obj = KF(F_,G_,H_,M_,Q_,R_,x0,P0)
            obj.F = F_;
            obj.G = G_;
            obj.H = H_;
            obj.M = M_;
            obj.Q = Q_;
            obj.R = R_;
            obj.x = x0;
            obj.P = P0;
        end
        
        function [x_curr,P_curr] = predict(obj,u)
            
            x_curr = obj.F*obj.x + obj.G*u;
            P_curr = obj.F*obj.P*obj.F' + obj.Q;
            
            obj.x = x_curr;
            obj.P = P_curr;
        end
        
        function [x_curr,P_curr] = update(obj,meas)
            for i=1:size(meas,2)
                
                % extract actual measurement
                meas_val = meas(:,i);
                
                % compute predicted measurement and innovation
                meas_pred = obj.H*obj.x;
                innov = meas_val - meas_pred;
                
                % compute Kalman gain
                K = obj.P*obj.H'/(obj.H*obj.P*obj.H' + obj.R);
                
                % update state
                x_curr = obj.x + K*innov;
                % update covariance
                P_curr = (eye(size(obj.F,1))-K*obj.H)*obj.P;
                
                obj.x = x_curr;
                obj.P = P_curr;
            end
        end
                   
    end
end

File: python/offset/filters/kf.py
#!/usr/bin/env python

__author__ = "Ian Loefgren"
__date__ = "1.7.2018"

"""
Linear Kalman filter implementation.
"""

import numpy as np

class KF(object):

    def __init__(self,F,G,H,M,Q,R_abs,R_rel,x0,P0,agent_id):
        self.F = F
        self.G = G
        self.H = H
        self.M = M
        self.Q = Q
        self.R_abs = R_abs
        self.R_rel = R_rel
        self.x = x0
        self.P = P0
        self.state_history = [x0]
        self.cov_history = [P0]
        self.agent_id = agent_id

    def predict(self,u):

        x_curr = np.dot(self.F,self.x) + np.dot(self.G,u)
        P_curr = np.dot(self.F,np.dot(self.P,self.F.transpose())) + self.Q

        self.x = x_curr
        self.P = P_curr

        self.state_history.append(x_curr)
        self.cov_history.append(P_curr)

        return x_curr, P_curr

    def update(self,meas,meas_type,src_id,rel_agent_id):

        for i in range(0,meas.size[0]):
            H = np.zeros([2,self.F.size[0]])
            if meas_type is 'abs':
                H[0,4*(src_id-1)] = 1
                H[1,4*(src_id-1)+2] = 1
                R = self.R_abs
            elif meas_type is 'rel':
                H[0,4*(src_id-1)] = 1
                H[1,4*(src_id-1)+2] = 1
                H[0,4*(rel_agent_id-1)] = -1
                H[1,4*(rel_agent_id-1)+2] = -1
                R = self.R_rel
            else:
                print('Received invalid measurement type {}'.format(meas_type))
                continue

            # extract measurement
            meas_val = meas[:,i:i+1]

            # compute predicted measurement and innovation
            meas_pred = np.dot(H,self.x)
            innov = meas_val - meas_pred

            # compute Kalman gain
            K = np.dot(np.dot(self.P,H.transpose()),np.linalg.inv(np.dot(np.dot(H,self.P),H.transpose()) + R))

            # update state
            x_curr = self.x + np.dot(K,innov)
            # update covariance
            P_curr = np.dot((np.eye(self.P.shape[0]) - np.dot(K,H)),self.P)

            self.x = x_curr
            self.P = P_curr
            self.state_history.append(x_curr)
            self.cov_history.append(P_curr)

            return x_curr,P_curr

def test_kf():
    pass

if __name__ == "__main__":
    test_kf()